cmd/noms/noms_root.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"strings"
+
+	"github.com/attic-labs/noms/cmd/util"
+	"github.com/attic-labs/noms/go/config"
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/hash"
+	flag "github.com/juju/gnuflag"
+)
+
+var nomsRoot = &util.Command{
+	Run:       runRoot,
+	UsageLine: "root <db-spec>",
+	Short:     "Get or set the current root hash of the entire database",
+	Long:      "See Spelling Objects at https://github.com/attic-labs/noms/blob/master/doc/spelling.md for details on the database argument.",
+	Flags:     setupRootFlags,
+	Nargs:     1,
+}
+
+var updateRoot = ""
+
+func setupRootFlags() *flag.FlagSet {
+	flagSet := flag.NewFlagSet("root", flag.ExitOnError)
+	flagSet.StringVar(&updateRoot, "update", "", "Replaces the entire database with the one with the given hash")
+	return flagSet
+}
+
+func runRoot(args []string) int {
+	if len(args) < 1 {
+		fmt.Fprintln(os.Stderr, "Not enough arguments")
+		return 0
+	}
+
+	cfg := config.NewResolver()
+	rt, err := cfg.GetRootTracker(args[0])
+	d.CheckErrorNoUsage(err)
+
+	currRoot := rt.Root()
+
+	if updateRoot == "" {
+		fmt.Println(currRoot)
+		return 0
+	}
+
+	fmt.Println(`WARNING
+
+This operation replaces the entire database with the instance having the given
+hash. The old database becomes eligible for GC.
+
+ANYTHING NOT SAVED WILL BE LOST
+
+Continue?
+`)
+	var input string
+	n, err := fmt.Scanln(&input)
+	d.CheckErrorNoUsage(err)
+	if n != 1 || strings.ToLower(input) != "y" {
+		return 0
+	}
+
+	if updateRoot[0] == '#' {
+		updateRoot = updateRoot[1:]
+	}
+	h, ok := hash.MaybeParse(updateRoot)
+	if !ok {
+		fmt.Fprintf(os.Stderr, "Error: Hash %s does not exist in database\n", h.String())
```

Comment:
`MaybeParse()` doesn't check whether the hash exists, it just determines whether the string parses and decodes to a byteslice of the right length.

If you're not going to validate anything about the object the provided hash points to, then change this error message to be something like "Hash %s does not parse".

If you _are_ going to validate whether this hash points to an existing object, you should probably also validate that that object has the appropriate type. For that you probably need a `Database`